chore(tools): Add `check` script

zharinov opened the pull request:
## Changes

Add `pnpm check` - a fast local CI script.

**Features:**
- Runs eslint-fix and prettier-fix first, then all other checks in parallel
- Only runs test shards affected by changed files (via `git diff` against base)
- Live terminal updates with spinners and progress timer
- Shows coverage gaps for changed files (only for full test runs, to avoid false positives)
- Graceful SIGINT handling with partial results

**Options:**
- `--no-fix` - Skip auto-fix (eslint-fix, prettier-fix)
- `--no-test` - Skip tests
- `--full` - Run full vitest instead of only affected shards
- `--base <branch>` - Compare against different base (default: main)

```
❯ pnpm check
Checks:
  eslint-fix ......................... ✓   2s
  prettier-fix ....................... ✓   2s
  ls-lint ............................ ✓   1s
  git-check .......................... ✓   1s
  markdown-lint ...................... ✓   3s
  doc-fence-check .................... ✓   5s
  lint-documentation ................. ✓   7s
  lint-other ......................... ✓   2s
  test-schema ........................ ✓   6s
  type-check ......................... ✓  20s

Total: 24s
```

## Context

- [x] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

- [x] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation).

## Documentation (please check one with an [x])

- [x] I have updated the documentation

## How I've tested my work (please select one)

- [x] Code inspection only

RahulGautamSingh commented on .github/contributing.md:
````diff
@@ -94,6 +94,16 @@ And some options:
+pnpm local-ci
+```
+
+This runs all checks in parallel: auto-fixes, lint, type-check, and only the test shards affected by your changes. Use `--no-tests` for lint-only, or `--no-fix` to skip auto-fixing.
````
PR description says `--fast` for not tests ?

zharinov replied:
Changed my mind: disable tests with --no-test, disable fix with --no-fix

RahulGautamSingh commented on .github/contributing.md:
````diff
@@ -94,6 +94,16 @@ And some options:
+For a fast (~1-2 min) local check that mirrors CI:
+
+```bash
+pnpm local-ci
````
```suggestion
pnpm check
```

viceice commented on tools/check.ts:
```diff
@@ -0,0 +1,709 @@
+ */
+
+import type { ChildProcess } from 'node:child_process';
+import { execSync, spawn } from 'node:child_process';
```
we should use `nano-spawn` or `execa` for better cross platform support

viceice commented on tools/check.ts:
```diff
@@ -0,0 +1,709 @@
+
+import type { ChildProcess } from 'node:child_process';
+import { execSync, spawn } from 'node:child_process';
+import { existsSync, readFileSync } from 'node:fs';
```
use `fs-extra` for better cross platform support and async io

viceice commented on tools/check.ts:
```diff
@@ -0,0 +1,709 @@
+): Promise<{ success: boolean; output: string }> {
+  return new Promise((resolve) => {
+    const proc = spawn(cmd, args, {
+      stdio: ['ignore', 'pipe', 'pipe'],
```
nano-spawn or execa makes std io handling much easier

RahulGautamSingh reviewed (changes requested):
```
for (const setting of pnpmWorkspace.minimumReleaseAgeExclude ?? []) 
```

The coverage report misses such cases ie. branched coverage, is it possible to handle?

zharinov committed:
refactor(tools): Split check.ts into modular components

- Separate files for types, coverage, git, tui, execution, test-filter
- Fix signal handler timing (register at module load)
- Export getMatchingShards/getTestPatternsForShards from test-shards.ts

zharinov commented on tools/check.ts:
```diff
@@ -0,0 +1,716 @@
+  duration: number,
+): string {
+  const nameCol = `${name} `.padEnd(CHECK_NAME_WIDTH, '.');
+  const status = success ? '✓' : '✗';
```
Fixed:

<img width="396" height="320" alt="image" src="https://github.com/user-attachments/assets/1f58e4d3-e03d-47f2-8d81-cd711cbb58eb" />

viceice commented on tools/check.ts:
```diff
@@ -0,0 +1,655 @@
+async function getChangedFiles(base: string): Promise<string[]> {
+  const files = new Set<string>();
+
+  async function addFromGitCommand(args: string[]): Promise<void> {
```
can we avoid nested functions?

viceice reviewed (commented):
we should split that file in future, it's pretty hard to follow 😅

zharinov committed:
base: pile foundation
